codalab/lib/codalab_manager_state.py

```diff
@@ -0,0 +1,201 @@
+        raise NotImplementedError
+
+    def set_auth(
+        self, server, access_token, expires_at, refresh_token, scope, token_type, username
+    ):
+        raise NotImplementedError
+
+    def delete_auth(self, server):
+        raise NotImplementedError
+
+    def get_session(self, name, default={}):
+        raise NotImplementedError
+
+    def set_session(self, name, address, worksheet_uuid):
+        raise NotImplementedError
+
+    def get_last_check_version_datetime(self, default=None):
+        raise NotImplementedError
+
+    def set_last_check_version_datetime(self, timestamp):
+        raise NotImplementedError
+
+
+class CodaLabManagerJsonState(CodaLabManagerState):
+    def initialize_state(self):
+        if self.temporary:
+            self.state = {'auth': {}, 'sessions': {}}
+            return
+        # Read state file, creating if it doesn't exist.
+        if not os.path.exists(self.state_path):
+            write_pretty_json(
+                {
+                    'auth': {},  # address -> {username, auth_token}
+                    'sessions': {},  # session_name -> {address, worksheet_uuid}
+                },
+                self.state_path,
+            )
+        self.state = read_json_or_die(self.state_path)
+
+    @property
+    def state_path(self):
+        return os.getenv('CODALAB_STATE', os.path.join(get_codalab_home(), 'state.json'))
+
+    def get_auth(self, server, default={}):
+        return deepcopy(self.state["auth"].get(server, default))
+
+    def set_auth(
+        self, server, access_token, expires_at, refresh_token, scope, token_type, username
+    ):
+        self.state["auth"][server] = {
+            "token_info": {
+                "access_token": access_token,
+                "expires_at": expires_at,
+                "refresh_token": refresh_token,
+                "scope": scope,
+                "token_type": token_type,
+            },
+            "username": username,
+        }
+        self._save_json_state()
+
+    def delete_auth(self, server):
+        self.state["auth"].pop(server)
+        self._save_json_state()
+
+    def get_session(self, name, default={}):
+        return deepcopy(self.state["sessions"].get(name, default))
+
+    def set_session(self, name, address, worksheet_uuid):
+        self.state["sessions"][name] = {"address": address, "worksheet_uuid": worksheet_uuid}
+        self._save_json_state()
+
+    def get_last_check_version_datetime(self, default=None):
+        return deepcopy(self.state.get("last_check_version_datetime", default))
+
+    def set_last_check_version_datetime(self, timestamp):
+        self.state["last_check_version_datetime"] = timestamp
+        self._save_json_state()
+
+    def _save_json_state(self):
+        if self.temporary:
+            return
+        write_pretty_json(self.state, self.state_path)
+
+
+class CodaLabManagerSqlite3State(CodaLabManagerState):
+    def initialize_state(self):
```

Comment:
Could you provide an high-level overview of the database schema / what's in the state?